libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1047,15 +1045,31 @@ string YulUtilFunctions::storageArrayPushZeroFunction(ArrayType const& _type)
 				if iszero(lt(oldLen, <maxArrayLength>)) { <panic>() }
 				sstore(array, add(oldLen, 1))
 				slot, offset := <indexAccess>(array, oldLen)
-				<storeValue>(slot, offset, <zeroValueFunction>())
+				<?isValueType>
+					<storeValue>(slot, offset, <zeroValueFunction>())
+				<!isValueType>
+					if gt(offset, 0) { invalid() }
+					let endData := add(slot, <storageSize>)
+					for { let currSlot := slot } lt(currSlot, endData) { currSlot := add(currSlot, 1) } {
+						sstore(currSlot, 0)
```

Comment:
Acutally I think this is not even needed - we usually clear when deleting, not when creating, so there should be zeros already.

Reply:
If that is true, than in case of push zero to storage array, we do not need to call this function at all?
It is than consider as a pure waste of `sstore`s? :)

Reply:
Indeed :)

Maybe we decided at some time to actually write zeros in the new codegen, but I cannot remember such a decision.

Reply:
Let's than leave as open, we can wait for @ekpyron opinion on this one. I am fine with any, but would prefer version that consumes less gas if it is safe :)

Reply:
It _should_ be safe :)

Reply:
Yeah, this is basically why I asked for properly resizing arrays before copying (which zeroes out the tail in case the array to be copied from is shorter than the one being copied to). Storage should always be kept zero, if unused.